### PR TITLE
Load and render parts of the gui at a later time

### PR DESCRIPTION
**Pull request by mzgoddard**
### Resolves

Load GUI and projects faster.

### Proposed Changes

Add a system to load and render parts of the GUI when JS seems to be idle or when those parts of the app must be rendered.

Dynamically load and render:
- Blocks
- MenuBar
- TargetPane
- CostumeTab
- SoundTab

### Reason for Changes

Scratch and its GUI have a set of work that can be performed in the background simultaneously with javascript. Network requests like those for project data and assets can be loaded while parts of GUI are rendered.

To support this goal work that isn't needed to make such network requests or other asynchronous activity should be delayed until it is needed or after that work when no other work is being performed. Work like fetching the project data leads into loading assets. So it also preferable that we perform any work loading assets will need before then.

### Benchmark Data

|  device | project | before (ms) | after (ms) | difference | change |
| --- | :--- | --- | --- | --- | --- |
|  chrome | default | 1297 | 1088 | 209 | 16.10% |
|  chrome | 173918262 | 2785 | 2348 | 437 | 15.69% |
|  safari | default | 1438 | 967 | 471 | 32.76% |
|  safari | 173918262 | 2847 | 2095 | 751 | 26.39% |
|  firefox  | default | 1569 | 1503 | 66 | 4.18% |
|  firefox  | 173918262 | 3239 | 2478 | 761 | 23.50% |
|  book | default | 3643 | 3014 | 629 | 17.28% |
|  book | 173918262 | 8404 | 7334 | 1069 | 12.73% |
|  ipad mini | default | 3140 | 2611 | 529 | 16.85% |
|  ipad mini | 173918262 | 7661 | 5935 | 1726 | 22.53% |
|  pi | default | 22916 | 20794 | 2122 | 9.26% |
|  pi | 173918262 | 58649 | 45655 | 12993 | 22.15% |

<details>
<summary>Full Data</summary>

|  device | project | branch | run 1 | run 2 | run 3 | run 4 | run 5 | average | stdev | difference | change |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
|  chrome | default | stagef | 888 | 872 | 935 | 914 | 881 | 898 | 26 | 399 | 30.76% |
|  chrome | default | gui | 1221 | 1078 | 1028 | 1118 | 996 | 1088 | 88 | 209 | 16.10% |
|  chrome | default | develop | 1661 | 1246 | 1194 | 1222 | 1162 | 1297 | 206 | 0 | 0.00% |
|  chrome | 173918262 | stagef | 2159 | 2243 | 2239 | 2281 | 2266 | 2238 | 47 | 547 | 19.64% |
|  chrome | 173918262 | gui | 2260 | 2592 | 2292 | 2294 | 2300 | 2348 | 138 | 437 | 15.69% |
|  chrome | 173918262 | develop | 2995 | 2640 | 2758 | 2842 | 2688 | 2785 | 140 | 0 | 0.00% |
|  book | default | stagef | 2685 | 2914 | 2688 | 2542 | 2562 | 2678 | 148 | 965 | 26.49% |
|  book | default | gui | 2876 | 2919 | 3106 | 2973 | 3196 | 3014 | 134 | 629 | 17.28% |
|  book | default | develop | 3658 | 3872 | 3476 | 3476 | 3735 | 3643 | 171 | 0 | 0.00% |
|  book | 173918262 | stagef | 6915 | 6272 | 6750 | 6287 | 6491 | 6543 | 284 | 1861 | 22.14% |
|  book | 173918262 | gui | 6646 | 8131 | 6578 | 6646 | 8670 | 7334 | 992 | 1069 | 12.73% |
|  book | 173918262 | develop | 8331 | 8505 | 8730 | 8177 | 8275 | 8404 | 218 | 0 | 0.00% |
|  safari | default | stagef | 816 | 1030 | 695 | 669 | 626 | 767 | 163 | 671 | 46.66% |
|  safari | default | gui | 1082 | 1025 | 1029 | 1010 | 690 | 967 | 157 | 471 | 32.76% |
|  safari | default | develop | 1665 | 1608 | 1131 | 1393 | 1395 | 1438 | 211 | 0 | 0.00% |
|  safari | 173918262 | stagef | 1671 | 1586 | 1694 | 1586 | 1686 | 1645 | 54 | 1202 | 42.23% |
|  safari | 173918262 | gui | 2399 | 2167 | 2102 | 1727 | 2082 | 2095 | 241 | 751 | 26.39% |
|  safari | 173918262 | develop | 2676 | 3125 | 2710 | 2996 | 2726 | 2847 | 201 | 0 | 0.00% |
|  firefox | default | stage | 1150 | 1449 | 1199 | 1635 | 1619 | 1410 | 228 | 158 | 10.10% |
|  firefox | default | gui | 1351 | 1767 | 1242 | 1162 | 1994 | 1503 | 360 | 66 | 4.18% |
|  firefox | default | develop | 1606 | 1599 | 1525 | 1599 | 1515 | 1569 | 45 | 0 | 0.00% |
|  firefox | 173918262 | stagef | 2329 | 2291 | 2126 | 2078 | 2104 | 2186 | 116 | 1054 | 32.53% |
|  firefox | 173918262 | gui | 2192 | 2301 | 2494 | 2985 | 2418 | 2478 | 306 | 761 | 23.50% |
|  firefox | 173918262 | develop | 3412 | 3512 | 3068 | 3023 | 3181 | 3239 | 214 | 0 | 0.00% |
|  ipad mini 10 | default | stagef | 2588 | 1727 | 1733 | 1734 | 1856 | 1928 | 373 | 1213 | 38.62% |
|  ipad mini 10 | default | gui | 2959 | 2137 | 2220 | 2852 | 2888 | 2611 | 398 | 529 | 16.85% |
|  ipad mini 10 | default | develop | 3840 | 3032 | 2866 | 2977 | 2987 | 3140 | 396 | 0 | 0.00% |
|  ipad mini 10 | 173918262 | stagef | 5941 | 6002 | 5269 | 5977 | 5723 | 5782 | 308 | 1878 | 24.52% |
|  ipad mini 10 | 173918262 | gui | 6326 | 5750 | 6053 | 5863 | 5683 | 5935 | 260 | 1726 | 22.53% |
|  ipad mini 10 | 173918262 | develop | 6889 | 7953 | 7849 | 7684 | 7929 | 7661 | 444 | 0 | 0.00% |
|  pi | default | stagef | 17080 | 16768 | 17214 | 17413 | 17284 | 17152 | 246 | 5764 | 25.15% |
|  pi | default | gui | 21239 | 21482 | 19854 | 21121 | 20274 | 20794 | 695 | 2122 | 9.26% |
|  pi | default | develop | 25202 | 22514 | 22103 | 21915 | 22846 | 22916 | 1328 | 0 | 0.00% |
|  pi | 173918262 | stagef | 47970 | 44499 | 47557 | 47974 | 45426 | 46685 | 1615 | 11963 | 20.40% |
|  pi | 173918262 | gui | 48674 | 43913 | 45708 | 45063 | 44918 | 45655 | 1806 | 12993 | 22.15% |
|  pi | 173918262 | develop | 59585 | 57628 | 57955 | 60683 | 57392 | 58649 | 1425 | 0 | 0.00% |

</details>

### CODE REVIEW

**Comment by mzgoddard** [comment]
This PR is architecturally and organizationally ready for review. I'm working on cleaning up remaining lint/test/doc issues.

**Comment by mzgoddard** [comment]
- Fix lint

File: src/lib/dynamic-render/selectors.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import {
+    getIsFetchingWithId,
+    getIsLoadingWithId,
+    LoadingState
+} from '../../reducers/project-state';
+
+// Selectors here can provide a descriptive interface for when delay arguments
+// should be which values. If we use only these common functions we can use that
+// as a way to shortcut all of the delays gates. By replacing the functions on
+// Delay with ones that return true.
```
Comment (paulkaplan) [inline src/lib/dynamic-render/selectors.js]: Hmmmm? I think there might be some punctuation out of place or something, but these sentences kinda frazzled my brain.

Comment (mzgoddard) [reply src/lib/dynamic-render/selectors.js]: Ah. Erm. Probably more than punctuation. Hmm.

**Comment by paulkaplan** [comment]
Closing this for now, since we made progress with a smaller version. We may want to revisit this in the future as a model for further splitting up the load

**Pull request closed without merging**